### ISSUE TITLE
Custom function not loading

### ISSUE THREAD

**Issue by victorq4** [OfficeDev/Excel-Custom-Functions#367]
*** NOTE: If you have an issue running your custom function, please log an issue in the official office-js repo here: [https://github.com/OfficeDev/office-js/issues](https://github.com/OfficeDev/office-js/issues) *** [This page takes me to a 404 not found page]

I have created an add-in using react. I used the yo office command to create it and then modified it to use custom functions and a taskpane. I implemented the shared runtime and the taskpane loads and works. The custom functions do not load anymore though. They were working fine until I had to clear office's cache. After that, they never loaded again. It does not show as an option when I type "=" on a cell and try to find it. My manifest passes the Manifest Validator test and everything seems to be correct. I have tried all troubleshooting suggestions online and nothing makes my custom functions load. I've tried associating them even though my .json file is generated automatically and I didn't have to do that before. I tried everything online but nothing seems to work. I feel the experience building add-ins is very poor. It's a pain to debug, the boiler plate for react should have an option to give you custom functions and a taskpane at once, debugging should be more convenient, etc. These just cause more trouble for Microsoft itself because they made it so boring and complicated that we have to bother them for help.

02/05/2023 11:42:44	Medium	Web Add-In Runtime Logging Session Started			
02/05/2023 11:42:44	Unexpected	Resource	The resource you are trying to use does not exist.		
02/05/2023 11:42:44	Medium	Resource	Resource Download Error Details: SolutionId=5fe02c13-6a6c-461d-9fc4-ab29b5922508 , AppVersion=1.0.0.0		
02/05/2023 11:42:44	Unexpected	Resource	The resource you are trying to use does not exist.		
02/05/2023 11:42:44	Verbose	CustomFunctions	[Metadata] [Parsing] [Begin]		
02/05/2023 11:42:44	Medium	CustomFunctions	[Metadata] [Parsing] [End] [Success]		
02/05/2023 11:42:44	Verbose	CustomFunctions	[Installation] [Addin] [Begin] Solution=wa104380862, Version=1.5.0.0		
02/05/2023 11:42:44	Medium	CustomFunctions	[Installation] [Addin] [End] [Success] Solution=wa104380862, Version=1.5.0.0		
02/05/2023 11:42:44	Verbose	CustomFunctions	[Installation] [Functions] [Begin] Solution=wa104380862, Workbook=Excel add-in 5fe02c13-6a6c-461d-9fc4-ab29b5922508.xlsx		
02/05/2023 11:42:44	Medium	CustomFunctions	[Installation] [Functions] [End] [Success] Solution=wa104380862, Workbook=Excel add-in 5fe02c13-6a6c-461d-9fc4-ab29b5922508.xlsx		
02/05/2023 11:42:45	Unexpected	Resource	The resource you are trying to use does not exist.		
02/05/2023 11:42:45	Unexpected	Resource	The resource you are trying to use does not exist.		

As you can see, the runtime logs seem to imply that my custom functions loaded correctly. Then it says a resource does not exist, without telling me which resource. How is that helpful? I checked all resources mentioned in my manifest and they do exist and are on the right path. 

[files.zip](https://github.com/OfficeDev/Excel-Custom-Functions/files/11372370/files.zip)

**Comment by AlexJerabek** [comment]
Hi @victorq4,

I'm sorry to hear you're having problems with Custom Functions. @yaweizhu-henson, can you please investigate?

I'm also going to tag @alison-mk, in case she has any tips to help.

**Comment by victorq4** [comment]
Hi team,
I am also having another problem with the authentication for my add-in.
Everything works fine until Auth0 sends the user to the redirect URI
provided. It is in the list of approved callback URLs on auth0 but the
dialog window crashes. I have tried changing the webpack configuration but
nothing seems to work.
[image: image.png]

Kind regards,
[image: Q4]

Victor Firmino

Analyst, Centre of Excellence - Q4 Inc.

M +44 (0) 7367 188904

L  +44 (0) 20 3950 2967
q4inc.com
q4blog.com
[image: Twitter]/q4tweets <https://twitter.com/q4tweets>
<https://q4marketing.s3.amazonaws.com/index.htm>
Revolutionize your IR program with Q4 Desktop
<https://www.q4inc.com/products/q4desktop/default.aspx>, your all-in-one
market intelligence and investor engagement platform. Visit www.q4inc.com to
learn more.
This email is intended only for the person or entity to which it is
addressed may contain confidential and/or privileged information. Any use
of this information by persons or entities other than the intended
recipient(s) is prohibited. If you received this in error, please contact
the sender and delete the email and all copies (electronic or otherwise)
immediately. Thank You.

**Comment by yaweizhu-henson** [comment]
Sorry for the late response. Our engineer has very limited bandwidth in CF area currently. I will try to push our engineer to have a look into this issue ASAP.

**Comment by donlvMSFT** [comment]
Hi @victorq4 , sorry to hear about the issues you've met.

I went through your manifest file, the definition went well as you mentioned.
As you mentioned they were working fine until I had to clear office's cache, have you npm install again after that?
Are you working on Excel online version or desktop version? If you could share your project without any private information, that would be helpful for us to locate the issue.

**Comment by victorq4** [comment]
[code.zip](https://github.com/OfficeDev/Excel-Custom-Functions/files/11442315/code.zip)
Here's a copy of my code without any private information. Thank you for looking into this. In the meantime, I have started a new project that does not use React and it's working so far. I haven't finished the migration yet though. I still need to work on the authentication bit. It'd be very helpful if the yeoman generator could create a react-based boiler plate with custom functions and a task pane on a shared runtime and also with the option to choose javascript or typescript.

**Comment by LandonFackrell** [comment]
I am currently developing an Excel Add-in with a coworker and we are having the same problem. I used the yeoman generator to generate the custom function boilerplate with a shared runtime and now I am unsure what happened but my functions don't appear in excel anymore. Is this issue still under review? Has there been any updates on possible fixes we can try?

**Comment by victorq4** [comment]
Hi Landon, so I could move on, I started over on a new project, without React. It worked for me and the functions appeared again and have been showing ever since. It was a hassle to move the code, but I couldn't figure out what was going on. I even tried to hire more professional developers to help me but they couldn't find the problem either, so it's probably a bug. They made it very complex and developer unfriendly. Even the documentation is hard to follow.

**Comment by victorq4** [comment]
Have you validated your manifest? It could be it. The add-in is very sensitive to the manifest. Run _npm run validate_ and confirm it says your manifest is valid. If there's anything I can help you with, reach out to me. I'm not the best at it, but I've been fighting my way through the challenges of developing this add-in for so long now that I'm not that bad either.

**Comment by LandonFackrell** [comment]
I totally agree with you this has been a monumental headache and very hard to follow even with the documentation. It looks like the manifest is valid. I am going to follow your lead and move the code over into a new project and see what I can do. Thanks so much @victorq4 ! You are a huge help!

**Comment by 1R053** [comment]
For me the basic "yo office" template for "Excel custom functions using a shared runtime" have that issue in both the JavaScript and the TypeScript version. I am using Excel on a Mac. The addins will start without errors, npm run validate is fine but the CONTOSO functions will not show up.

**Comment by randose** [comment]
Thanks for the quick response! I'm sorry to hear you got that frustrated with the office add-in, but sounds like google sheets is working great. I'm determined to figure this thing out for Excel because my use case requires it.

I found this on the Microsoft Learn docs that might explain a lot:
![image](https://github.com/OfficeDev/Excel-Custom-Functions/assets/5472716/5ab000c7-b0d5-4fc8-929f-0af7bdcccd64)

So I'm exploring alternate ways to test out the functions on Excel using the instructions here: [Debug add-ins using developer tools...](https://learn.microsoft.com/en-us/office/dev/add-ins/testing/debug-add-ins-using-devtools-edge-chromium)

Haven't gotten it to work yet though!

**Comment by 1R053** [comment]
@millerds this issue also occurs on a Mac with this repo. i.e. I download the branch here from github and run it with npm. The sidepar shows up, custom functions don't. I also tried cleaning the cache, but with that situation a cache issue should be ruled out I assume.

**Comment by victorq4** [comment]
Clearing the cache actually causes the function to disappear if it was showing before. At least that's what would always happen to me.

**Comment by randose** [comment]
I got it working on Excel for Windows just now!

Before I explain my steps, please note that I created my project using the Yeoman Generator for Office, with the React template.  This may differ from how you set up your project, and as such, my right answers might be wrong answers for you - but they might still help get you going in the right direction.
By default, the Yeoman Office React template does not include Custom Functions, nor does it include a Shared Runtime, the latter of which, as I understand it, is required to run Custom Functions in new versions of Excel for Windows. So, I modified my webpack.config.js and manifest.xml files to support a Shared Runtime and Custom Functions. I also had to install a few missing node modules. To do this, I used "yo office" again and created a new project with the "Custom Functions with Shared Runtime" template. I compared my webpack.config.js, manifest.xml, and package.json files to figure out what was missing, and then added and/or modified the missing pieces.
I go over these modifications in steps 4-6 below.

After I went through those steps, I made my first frustrated comment above. Since then, I've resolved the issue.
Here were my steps, in reverse order. Hopefully this can help one / all of you frustrated individuals that I feel so deeply for:

1. For me, the final issue was that I missed the basic but important step of going to the "Insert" tab on the Ribbon, selecting the dropdown next to My "Add-Ins", and then clicking on my app in the Developer Add-Ins section. Doing this activates the custom functions! This process is detailed in this helpful guide: [Custom functions quick start - Office Add-ins | Microsoft Learn](https://learn.microsoft.com/en-us/office/dev/add-ins/quickstarts/excel-custom-functions-quickstart?view=powerpoint-js-preview&tabs=excel-windows#try-out-a-prebuilt-custom-function). If you don't see your app there, you may have another issue and should read on.
2. I was also getting the issue I described in my earlier post, where upon loading Excel from "npm start", there was an error in Excel's bottom status bar saying something to the effect of "Error loading add-in: Some resources of an add-in couldn't be loaded". It turns out that I had given the wrong URL for "Functions.Script.Url" and "Functions.Metadata.Url" in my manifest.xml file. For me, the correct URLs were "https://localhost:3000/functions.js" & "https://localhost:3000/functions.json", respectively. Previously, I was using "https://localhost:3000/functions/functions.js" (and the equivalent for "functions.json") - this is how my src/ folder is structured, and this is how I saw it done elsewhere (see step 6 below), however I realized that webpack actually minifies everything at runtime into one flat directory, meaning functions.js and functions.json would be at the root of localhost:3000. This can be observed by running "npm run build", and then looking through the dist/ folder, which is flat and should contain functions.js and functions.json. If, after running "npm run build", your dist/ folder doesn't contain those files, read on.
3. It goes without saying that your function code in functions.js should be correct, and have the proper code commenting. The code commenting allows webpack to automatically create the functions.json file for you at compile time. There are some decent instructions here on how to write custom functions code with the commenting tags: [Custom Functions Overview](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-overview)
I used a very simple Hello World function for troubleshooting purposes. The entirety of my functions.js file is below:
```
/* global CustomFunctions */

/**
 * Hello World function
 * @customfunction HW
 * @param [name] The name to use in the greeting.
 * @returns The string "Hello World!".
 */
function HelloWorld(name = null) {
  return `Hello ${name || "World"}!`;
}

CustomFunctions.associate("HW", HelloWorld);
```
4. Make sure you have the following node modules installed as devDependencies in package.json: "@types/custom-functions-runtime", and "custom-functions-metadata-plugin". If not, install them with "npm install --save-dev "PACKAGE_NAME" (replace "PACKAGE_NAME" with the packages I just mentioned).
5. Make sure your webpack.config.js file has the following lines:
Toward the top with the other require statements:
`const CustomFunctionsMetadataPlugin = require("custom-functions-metadata-plugin");`

After urlDev and urlProd definitions, enter this to appease ESLint/Prettier:
`/* global require, module, process */`

In the definition of config inside the module.exports definition:
```
entry: {
  ...
  functions: "./src/functions/functions.js",
},
...
```

Also inside the config object:
```
...
plugins: [
  new CustomFunctionsMetadataPlugin({
    output: "functions.json",
    input: "./src/functions/functions.js",
  }),
...
```
Finally, there should only be ONE HtmlWebpackPlugin instantiation under the plugins section of config (referenced just above this), and it should look something like this: 
```
...,
new HtmlWebpackPlugin({
  filename: "taskpane.html",
  template: "./src/taskpane/taskpane.html",
  chunks: ["taskpane", "vendor", "polyfill", "commands", "functions"],
}),
...
```
(Note that in the above, if your app is not using React, the "vendor" chunk may not be listed and I think that is okay)

7. Make sure your manifest.xml file has all of the relevant entries to define a Shared Runtime and Custom Functions. This is the example that I went off of, but keep in mind that it led me astray as to the Functions.Script.Url and Functions.Metadata.Url entries, which I explained and corrected in Step 2 above: [OfficeDev Sample with Shared Runtime](https://github.com/OfficeDev/Office-Add-in-samples/blob/main/Samples/excel-shared-runtime-global-state/manifest-localhost.xml)

If this helps one of you get it working, I'll be happy. Best of luck to all!

**Comment by Rick-Kirkham** [comment]
It looks like a @randose has found a fix/workaround, so I will close this issue. If anyone thinks it has been closed in error, please add a comment to that effect.
